src/GitHub.App/Services/NavigationService.cs

```diff
@@ -0,0 +1,123 @@
+        {
+            int line;
+            int column;
+            ErrorHandler.ThrowOnFailure(sourceView.GetCaretPos(out line, out column));
+            var text1 = GetText(sourceView);
+
+            var view = OpenDocument(targetFile);
+            var text2 = VsShellUtilities.GetRunningDocumentContents(serviceProvider, targetFile);
+
+            var fromLines = ReadLines(text1);
+            var toLines = ReadLines(text2);
+            var matchingLine = FindNearestMatchingLine(fromLines, toLines, line);
+            if (matchingLine == -1)
+            {
+                // If we can't match line use orignal as best guess.
+                matchingLine = line < toLines.Count ? line : toLines.Count - 1;
+                column = 0;
+            }
+
+            ErrorHandler.ThrowOnFailure(view.SetCaretPos(matchingLine, column));
+            ErrorHandler.ThrowOnFailure(view.CenterLines(matchingLine, 1));
+
+            return view;
+        }
+
+        public IVsTextView FindActiveView()
+        {
+            var textManager = (IVsTextManager2)serviceProvider.GetService(typeof(SVsTextManager));
+            IVsTextView view;
+            var hresult = textManager.GetActiveView2(1, null, (uint)_VIEWFRAMETYPE.vftCodeWindow, out view);
+            return hresult == VSConstants.S_OK ? view : null;
+        }
+
+        string GetText(IVsTextView textView)
+        {
+            IVsTextLines buffer;
+            ErrorHandler.ThrowOnFailure(textView.GetBuffer(out buffer));
+
+            int line;
+            int index;
+            ErrorHandler.ThrowOnFailure(buffer.GetLastLineIndex(out line, out index));
+
+            string text;
+            ErrorHandler.ThrowOnFailure(buffer.GetLineText(0, 0, line, index, out text));
+            return text;
+        }
+
+        IVsTextView OpenDocument(string fullPath)
+        {
+            var logicalView = VSConstants.LOGVIEWID.TextView_guid;
+            IVsUIHierarchy hierarchy;
+            uint itemID;
+            IVsWindowFrame windowFrame;
+            IVsTextView view;
+            VsShellUtilities.OpenDocument(serviceProvider, fullPath, logicalView, out hierarchy, out itemID, out windowFrame, out view);
+            return view;
+        }
+
+        static int FindNearestMatchingLine(IList<string> fromLines, IList<string> toLines, int line)
+        {
+            line = line < fromLines.Count ? line : fromLines.Count - 1; // VS shows one extra line at end
+            var fromLine = fromLines[line];
+
+            for (var offset = 0; true; offset++)
+            {
+                var lineAbove = line + offset;
+                var checkAbove = lineAbove < toLines.Count;
+                if (checkAbove && toLines[lineAbove] == fromLine)
+                {
+                    return lineAbove;
+                }
+
+                var lineBelow = line - offset;
+                var checkBelow = lineBelow >= 0;
```

Comment:
This `checkBelow` doesn't match the `checkAbove` (which seems to be making sure we don't do an out of bounds array access), it probably should match it?

Reply:
This simple search algorithm searches outwards from the line number it expects to find a match on. The `checkBelow` and `checkAbove` flags indicate whether there are still lined above or below to search. When there are no more lines above or below to search it gives up.

Although it partially uses `lineBelow` for its bounds check, it also includes `&& lineBelow < toLines.Count`. It doesn't want to abort searching simply because it starts off screen (which would happen if `lineBelow` included this check).